Sources/NIO/ByteBuffer-core.swift

```diff
@@ -492,14 +492,32 @@ public struct ByteBuffer {
         return try body(.init(rebasing: self._slicedStorageBuffer.dropFirst(self.writerIndex)))
     }
 
+    /// This vends a pointer of the `ByteBuffer` at the `writerIndex` after ensuring that the buffer has at least `minimumWritableBytes` of writable bytes available.
+    ///
+    /// - warning: Do not escape the pointer from the closure for later use.
+    ///
+    /// - parameters:
+    ///     - minimumWritableBytes: The number of writable bytes to reserve capacity for before vending the `ByteBuffer` pointer to `body`.
+    ///     - body: The closure that will accept the yielded bytes and return the number of bytes written.
+    /// - returns: The number of bytes written.
     @discardableResult
     @inlinable
-    public mutating func writeWithUnsafeMutableBytes(_ body: (UnsafeMutableRawBufferPointer) throws -> Int) rethrows -> Int {
+    public mutating func writeWithUnsafeMutableBytes(minimumWritableBytes: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> Int) rethrows -> Int {
+        if minimumWritableBytes > 0 {
+            self.reserveCapacity(writerIndex + minimumWritableBytes)
+        }
         let bytesWritten = try withUnsafeMutableWritableBytes(body)
         self._moveWriterIndex(to: self._writerIndex + _toIndex(bytesWritten))
         return bytesWritten
     }
 
+    @available(*, deprecated, message: "please use writeWithUnsafeMutableBytes(minimumWritableBytes:_:) instead to ensure sufficient write capacity.")
+    @discardableResult
+    @inlinable
+    public mutating func writeWithUnsafeMutableBytes(_ body: (UnsafeMutableRawBufferPointer) throws -> Int) rethrows -> Int {
```

Comment:
Not sure if we should really deprecate this function? It doesn't cause any harm. CC @Lukasa / @glbrntt

Reply:
I don't think we should: I think there are cases where you can elide the `self.reserveCapacity` call because enough space has been allocated ahead of time and the number of bytes written by `writeWithUnsafeMutableBytes` does not account for all of the bytes (i.e. allocate X bytes, write 4 bytes as an Int32, write X-4 bytes with `writeWithUnsafeMutableBytes`).

Reply:
Yeah, I tend I agree. @Lukasa asked me to deprecate it but as I was moving the codebase off of the now deprecated API, I was having to specify the ```minimumWritableBytes``` in places where the buffer had just been instantiated with the right size. So I was having to repeat myself unnecessarily.

Reply:
I think we should deprecate it because it is easier to misuse than anything else. It simply doesn't hurt to ask for the number of bytes you want, even if that's repeating yourself. It also avoids bugs by having NIO double-check your work. Specifically, `reserveCapacity` is an extremely cheap function call in the case where it does nothing, so I don't think the cost of making it hurts at all.

Reply:
I'm willing to do it either way, so just let me know what you guys decide.

Reply:
I'm still in favour of not deprecating the old version because I don't see the real point. But I'm not strong on it. So could me as

> weakly if favour of keeping the old version

@Lukasa are you strongly or weakly in favour of removing the old version
@glbrntt are you strongly or weakly in favour of keeping the old version (or did you change your mind)
@2bjake do you have any further thoughts or would you like to just defer to us?

Reply:
Strongly in favour of removing. I don't think the old version serves any value with the new one in existence. If we implemented this new version first we'd never write the old one, so we should move towards the better API.

Reply:
I'm weakly in favour of keeping the older version.

Reply:
Assuming @2bjake is not 'strongly in favour of keeping', let's kill it then (I'm counting weakly about as much as 0.3 or less)

Reply:
Sounds good.

Reply:
Yeah, I'm happy to defer to y'all.